### PR TITLE
apply_ufunc vectorize 1D function example

### PR DESCRIPTION
**Pull request by dcherian**
I added an example on using apply_ufunc to vectorize a 1D example over a DataArray. Comments and feedback welcome. 

I added an example of using numba's guvectorize too. Thoughts on keeping that bit?

 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

cc @rabernat @TomNicholas

### CODE REVIEW

**Comment by TomNicholas** [comment]
This is great! I really like the idea of guiding people through the various errors produced.

I'm going to try and get some more feedback on this from some of my labmates, but one suggestion is maybe explicitly demonstrate that now the function can be applied to an N-D array? i.e. apply it over longitude too? Maybe also show the signature of the final call at the start of the tutorial, so that readers know what you are moving towards?

(There are also a couple of small typos, I can point them out explicitly later if you want)

**Comment by dcherian** [comment]
> but one suggestion is maybe explicitly demonstrate that now the function can be applied to an N-D array? 

Are you referring to the `interp1d_gufunc` function?

> I'm going to try and get some more feedback on this from some of my labmates

Great! Looking forward to it.

Feel free to push to this branch. I've enabled "edits from maintainers" so you should be able to.

**Comment by dcherian** [comment]
ready for merge.

File: doc/examples.rst [hunk 1]
```diff
@@ -10,3 +10,10 @@ Examples
     examples/visualization_gallery
     examples/ROMS_ocean_model
     examples/ERA5-GRIB-example
+
+apply_ufunc
+------------
+.. toctree::
+    :maxdepth: 2
+
```
Comment (keewis) [inline doc/examples.rst]: nbsphinx automatically adds sections from the notebooks, so unless you want to add more `apply_ufunc` examples you might want to remove the explicit section to be consistent with the other notebooks.

Comment (dcherian) [reply doc/examples.rst]: >  unless you want to add more apply_ufunc examples

Yes, this is my intent.

Comment (keewis) [reply doc/examples.rst]: Then, would it be better to replace `apply_ufunc` with "Using `apply_ufunc`" or something like that?

File: doc/examples/apply_ufunc_vectorize_1d.ipynb [hunk 2]
```diff
@@ -0,0 +1,2762 @@
...
+      "\u001b[0;32m~/miniconda3/envs/dcpy/lib/python3.7/site-packages/xarray/core/computation.py\u001b[0m in \u001b[0;36mapply_ufunc\u001b[0;34m(func, input_core_dims, output_core_dims, exclude_dims, vectorize, join, dataset_join, dataset_fill_value, keep_attrs, kwargs, dask, output_dtypes, output_sizes, *args)\u001b[0m\n\u001b[1;32m    982\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mexclude_dims\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mexclude_dims\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0msignature\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_core_dims\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    983\u001b[0m         raise ValueError(\n\u001b[0;32m--> 984\u001b[0;31m             \u001b[0;34m\"each dimension in `exclude_dims` must also be a \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    985\u001b[0m             \u001b[0;34m\"core dimension in the function signature\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u00
```
Comment (keewis) [inline doc/examples/apply_ufunc_vectorize_1d.ipynb]: The first sentence looks like it was cut short and should have ended with something like "... provides some information on what is going on."

Also, nit: `print` is not a statement in py3 anymore so this should be "... but the output of our `print` call" or "... but our `print` call output"

File: doc/examples.rst (hunk 1 above)
Comment (dcherian) [reply doc/examples.rst]: :+1:

File: doc/examples/apply_ufunc_vectorize_1d.ipynb (hunk 2 above)
Comment (dcherian) [reply doc/examples/apply_ufunc_vectorize_1d.ipynb]: Fixed both. Thanks.

I've also stripped the output.

**Pull request merged**